### PR TITLE
Remove bit reverse from Parity Doughnut, signing is now valid

### PR DESCRIPTION
**Pull request by hoani**
A correction to LE encoding based on: https://github.com/cennznet/doughnut-paper/pull/9

LE encoding is applied at a byte level, but not at a bit level.

Relates to: https://github.com/cennznet/doughnut-maker/pull/15

## Changes

* Removed all bit-swaps
* Removed `unsafe` allocation on a vector in signature decoding
* Refactored a lot of tests due to incompatible signatures and encoding
* Add some common tests shared with `doughnut_maker` app

## Concerns

* The V0 doughnut really doesn't do anything, and its tests are getting frustrating to maintain - should remove it soon and just have the module as a thin wrapper for the Parity doughnut module.
* We are adding some coupling with the common tests

### CODE REVIEW

File: src/test.rs [hunk 1]
```diff
@@ -155,21 +179,26 @@ fn it_works_v0_parity_ed25519() {
...
+        191, 33, 202, 128, 138, 165, 73, 244, 67, 247, 37, 13, 218, 44, 244, 54, 137, 179, 56, 110,
+        152, 170, 180, 218, 107, 177, 170, 58, 91, 62, 24, 240, 248, 244, 13, 51, 235, 3, 21, 63,
+        79, 192, 137, 6,
     ];
     let doughnut = Doughnut::decode(&mut &payload[..]).expect("It works");
     let d = ParityDoughnutV0::try_from(doughnut.clone()).unwrap();
 
     assert_eq!(d.signature_version, 3);
-    assert_eq!(d.payload_version, 2);
+    assert_eq!(d.payload_version, 0);
```
Comment (holygits) [inline src/test.rs:201]: IIRC this was testing the encoding of payload version when non-zero
It should have it's own test but mentioning as an FYI incase

File: src/traits/impls.rs [hunk 2]
```diff
@@ -93,97 +93,131 @@ impl DoughnutVerify for Doughnut {
...
+        let header: Vec<u8> = vec![0, 8, 3];
+        let issuer = keypair.public.to_bytes().to_vec();
+        let holder = vec![0x15; 32];
+
+        let payload: Vec<u8> = [header, issuer, holder, test_domain_data()].concat();
+        let signature = keypair.sign(&payload);
+
+        let mut encoded: Vec<u8> = [payload, signature.to_bytes().to_vec()].concat();
+        let index = encoded.len() - 1;
+        encoded[index] = 0x00;
```
Comment (holygits) [inline src/traits/impls.rs:217]: optional: comment would be nice here

File: src/traits/impls.rs [hunk 3]
```diff
@@ -93,97 +93,131 @@ impl DoughnutVerify for Doughnut {
 #[cfg(test)]
 mod test {
     use super::*;
+    use crate::signature::CONTEXT_ID;
     use crate::traits::DoughnutVerify;
+    use bit_reverse::ParallelReverse;
```
[the remark this answers is not in the document]
Comment (hoani) [reply src/traits/impls.rs:98]: Yes, we can remove it from `src/traits/impls.rs`, we cannot remove it from deps unless we remove the `DoughnutV0` binary thing (which I mentioned I want to remove in the concerns, but I don't feel like this PR is the place)

File: src/test.rs (hunk 1 above)
Comment (hoani) [reply src/test.rs:201]: That's cool - I think with the latest encode/decode testing PR, we have sufficient coverage elsewhere

**Pull request merged**